blog/2016-04-20-futhark-and-pygame.rst

```diff
@@ -0,0 +1,208 @@
+Futhark development does *not* target interactive visualisations --
+it's just nice to show off what we have!
+
+
+A simple start: Game of Life
+----------------------------
+
+Let's start with Futhark's own Game of Life visualisation.  This is
+arguably the simplest of the four visualisations, as it's actually not
+interactive.  However, it shows how few lines of code are needed to
+integrate Futhark, NumPy, and PyGame.
+
+First, here's a video of me running the visualisation!
+
+.. raw:: html
+
+   <video src="/static/life-2016.04.20.webm"
+          poster="/static/life-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/life-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/life-2016.04.20.webm">life-2016.04.20.webm</a></p>
+
+Secondly, here's how it works: ``life-visualise.py`` first sets up PyGame::
+
+  screen = pygame.display.set_mode(size)
+
+where ``size`` is a tuple of ``(width, height)``.  We then generate a
+NumPy array of random initial states::
+
+  initworld = numpy.random.choice([True, False], size=size)
+
+A NumPy array can be used as an argument to an entry point function in
+a Futhark Python module.  In this case, we use the NumPy array to get
+the initial simulation values by calling::
+
+  world, history = l.init(initworld)
+
+We also need a temporary PyGame surface for transferring pixel data
+from NumPy to PyGame, so we run this::
+
+  surface = pygame.Surface(size)
+
+To render a frame, we use the temporary ``surface`` surface to
+transform the returned NumPy array into something PyGame can
+understand, after which we blit it to the screen::
+
+  def render():
+      frame = l.render_frame(history)
+      pygame.surfarray.blit_array(surface, frame)
+      screen.blit(surface, (0, 0))
+      pygame.display.flip()
+
+Here, the ``l`` variable is the imported simulation backend,
+i.e. either the default ``life`` or one of the alternative variants
+``quadlife`` or ``quadlife_alt``, whose rules are invented by Torben
+Mogensen.  When running the visualisation, you can specify the backend
+with the ``--variant`` command-line flag.
+  
+Finally, since Game of Life is a state-based simulation, we need a way
+to step through the simulation, using previous output as new input.
+This is pretty simple in Python::
+
+  while True:
+      world, history = l.steps(world, history, steps)
+      render()
+      for event in pygame.event.get():
+          if event.type == pygame.QUIT:
+              sys.exit()
+
+The ``steps`` argument is the number of simulation steps to perform
+per frame, and defaults to ``3``.  We have also added a simple PyGame
+event check, so that you can close the simulation window as expected.
```

Comment:
Somewhere we should mention why this is useful (to increase the amount of work done per frame).